pkg/securelaunch/measurement/cpuid.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/u-root/u-root/pkg/securelaunch/tpm"
+	"io"
+	"log"
+	"strings"
+)
+
+const (
+	defaultCPUIDFile = "cpuid.txt" //only used if user doesn't provide any
+)
+
+type CPUIDCollector struct {
+	Type     string `json:"type"`
+	Location string `json:"location"`
+}
+
+func NewCPUIDCollector(config []byte) (Collector, error) {
+	log.Printf("New CPUID Collector initialized\n")
+	var fc = new(CPUIDCollector)
+	err := json.Unmarshal(config, &fc)
+	if err != nil {
+		return nil, err
+	}
+	return fc, nil
+}
+
+func getCPUIDInfo() []byte {
+	var w strings.Builder
+	fmt.Fprintf(&w, "VendorString:           %s\n", cpuid.VendorIdentificatorString)
+	fmt.Fprintf(&w, "ProcessorBrandString:   %s\n", cpuid.ProcessorBrandString)
+	fmt.Fprintf(&w, "SteppingId:     %d\n", cpuid.SteppingId)
+	fmt.Fprintf(&w, "ProcessorType:  %d\n", cpuid.ProcessorType)
+	fmt.Fprintf(&w, "DisplayFamily:  %d\n", cpuid.DisplayFamily)
+	fmt.Fprintf(&w, "DisplayModel:   %d\n", cpuid.DisplayModel)
+	fmt.Fprintf(&w, "CacheLineSize:  %d\n", cpuid.CacheLineSize)
+	fmt.Fprintf(&w, "MaxLogocalCPUId:%d\n", cpuid.MaxLogocalCPUId)
+	fmt.Fprintf(&w, "InitialAPICId:  %d\n", cpuid.InitialAPICId)
+	fmt.Fprintf(&w, "Smallest monitor-line size in bytes:  %d\n", cpuid.MonLineSizeMin)
+	fmt.Fprintf(&w, "Largest monitor-line size in bytes:   %d\n", cpuid.MonLineSizeMax)
+	fmt.Fprintf(&w, "Monitor Interrupt break-event is supported:  %v\n", cpuid.MonitorIBE)
+	fmt.Fprintf(&w, "MONITOR/MWAIT extensions are supported:      %v\n", cpuid.MonitorEMX)
+	fmt.Fprintf(&w, "AVX state:     %v\n", cpuid.EnabledAVX)
+	fmt.Fprintf(&w, "AVX-512 state: %v\n", cpuid.EnabledAVX512)
+	fmt.Fprintf(&w, "Interrupt thresholds in digital thermal sensor: %v\n", cpuid.ThermalSensorInterruptThresholds)
+
+	fmt.Fprintf(&w, "Features: ")
+	for i := uint64(0); i < 64; i++ {
+		if cpuid.HasFeature(1 << i) {
+			fmt.Fprintf(&w, "%s ", cpuid.FeatureNames[1<<i])
+		}
+	}
+	fmt.Fprintf(&w, "\n")
+
+	fmt.Fprintf(&w, "ExtendedFeatures: ")
+	for i := uint64(0); i < 64; i++ {
+		if cpuid.HasExtendedFeature(1 << i) {
+			fmt.Fprintf(&w, "%s ", cpuid.ExtendedFeatureNames[1<<i])
+		}
+	}
+	fmt.Fprintf(&w, "\n")
+
+	fmt.Fprintf(&w, "ExtraFeatures: ")
+	for i := uint64(0); i < 64; i++ {
+		if cpuid.HasExtraFeature(1 << i) {
+			fmt.Fprintf(&w, "%s ", cpuid.ExtraFeatureNames[1<<i])
+		}
+	}
+	fmt.Fprintf(&w, "\n")
+
+	fmt.Fprintf(&w, "ThermalAndPowerFeatures: ")
+	for i := uint32(0); i < 64; i++ {
+		if cpuid.HasThermalAndPowerFeature(1 << i) {
+			if name, found := cpuid.ThermalAndPowerFeatureNames[1<<i]; found {
+				fmt.Fprintf(&w, "%s ", name)
+			}
+		}
+	}
+	fmt.Fprintf(&w, "\n")
+
+	for _, cacheDescription := range cpuid.CacheDescriptors {
+		fmt.Fprintf(&w, "CacheDescriptor: %v\n", cacheDescription)
+	}
+	return []byte(w.String())
+}
+
+func measureCPUIDFile(tpmHandle io.ReadWriteCloser) ([]byte, error) {
+
+	d := getCPUIDInfo() // return strings builder
+	if e := tpm.ExtendPCRDebug(tpmHandle, pcr, d); e != nil {
```

Comment:
I'm curious about this... rather than a structured format, you PCR-extend a string here. 

Doesn't that mean that if the format of this string evolves, you'll have to adjust your expected PCR values based on potentially both possible string formats? (Think deployment -- half the fleet on the old u-root, half the fleet on the new u-root, and your attestation product now has to support both formats, which means most likely an old vendored version of this and a new one?)

Or am I on the wrong path here?

Reply:
Truth is this was a last minute requirement and we made it pull a core set of attributes. In the next release we will expand it to function like the DMI measurement where the policy will dictate what cpuid fields to include into the measurement. Like DMI this is a description of the hardware present, as long as we are predictable in how we represent each flag and ordering then an attestation service can review the policy file and predict the value using inventory data.

With that said, what is worth consideration is the number of bytes to hash when contrasting a readable string representation and a binary structure.

Reply:
I can kind of already see the future in which you want to commit a copy of this code with additional information without deleting the old version, because you need both the old and the new one to be around for the attestation service to know how to predict the data. Wouldn't that be easier to manage if the data was actually structured with something like json? (I don't think it has to be a binary structure either.)

Then the service could just keep a mapping of version number to field-names expected to be correct.

Reply:
(I'm okay committing this as is -- let me know what you want to do.)

Reply:
okay so we decided to go with intel-go/cpuid package since it has "Stepping" information.